runtime/service_integ_test.go

```diff
@@ -1104,3 +1104,91 @@ func TestUpdateVMMetadata_Isolated(t *testing.T) {
+	defer client.Close()
+
+	image, err := alpineImage(ctx, client, defaultSnapshotterName())
+	require.NoError(t, err, "failed to get alpine image")
+	containerName := "test-entropy"
+
+	randomBlockCount := "1024"
+	ddContainer, err := client.NewContainer(ctx,
+		containerName,
+		containerd.WithSnapshotter(defaultSnapshotterName()),
+		containerd.WithNewSnapshot("test-entropy-snapshot", image),
+		containerd.WithNewSpec(
+			oci.WithDefaultUnixDevices,
+			// Use blocksize of 2500 as rngtest consumes data in blocks of 2500 bytes.
+			oci.WithProcessArgs("/bin/dd", "if=/dev/random", "of=/dev/stdout", "bs=2500", "count="+randomBlockCount),
+		),
+	)
+	require.NoError(t, err, "failed to create container %s", containerName)
+
+	// rngtest is a utility to "check the randomness of data using FIPS 140-2 tests", installed as part of
+	// the container image this test is running in. We pipe the output from "dd if=/dev/random" to rngtest
+	// to validate the quality of the randomness inside the VM.
+	// TODO It would be conceptually simpler to just run rngtest inside the container in the VM, but
+	// doing so would require some updates to our test infrastructure to support custom-built container
+	// images running in VMs (right now it's only feasible to use publicly available container images).
+	// Right now, it's instead run as a subprocess of this test outside the VM.
+	var rngtestStdout bytes.Buffer
+	var rngtestStderr bytes.Buffer
+	rngtestCmd := exec.CommandContext(ctx, "rngtest",
+		"--blockcount="+randomBlockCount,
+	)
+	rngtestCmd.Stdout = &rngtestStdout
+	rngtestCmd.Stderr = &rngtestStderr
+	rngtestStdin, err := rngtestCmd.StdinPipe()
+	require.NoError(t, err, "failed to get pipe to rngtest command's stdin")
+
+	ddStdout := rngtestStdin
+	var ddStderr bytes.Buffer
+
+	task, err := ddContainer.NewTask(ctx, cio.NewCreator(cio.WithStreams(nil, ddStdout, &ddStderr)))
+	require.NoError(t, err, "failed to create task for dd container")
+
+	exitCh, err := task.Wait(ctx)
+	require.NoError(t, err, "failed to wait on task for dd container")
+
+	err = task.Start(ctx)
+	require.NoError(t, err, "failed to start task for dd container")
+
+	err = rngtestCmd.Start()
+	require.NoError(t, err, "failed to start rngtest")
+
+	select {
+	case exitStatus := <-exitCh:
+		assert.NoError(t, exitStatus.Error(), "failed to retrieve exitStatus")
+		assert.Equal(t, uint32(0), exitStatus.ExitCode())
+
+		status, err := task.Delete(ctx)
+		assert.NoErrorf(t, err, "failed to delete dd task after exit")
+		if status != nil {
+			assert.NoError(t, status.Error())
+		}
+
+		t.Logf("stderr output from dd:\n %s", ddStderr.String())
+	case <-ctx.Done():
+		require.Fail(t, "context cancelled",
+			"context cancelled while waiting for dd container to exit (is it blocked on reading /dev/random?), err: %v", ctx.Err())
+	}
+	rngtestStdin.Close()
+
+	// rngtest will exit non-zero if any blocks fail its randomness tests
```

Comment:
If I get this properly, rngtest will return `1 if no errors happen, but at least one block fails the FIPS tests`. Can this potentially be a flaky test that we'll have to retry? May be it worth to define some success threshold, like no more than 1% or something.

Reply:
Good question, I ran a test where I locally increased the number of blocks by a factor of 1024 and got 27 block failures out of 32863 total tested. So, presuming those are all false-negatives and that I'm remembering how to math, if we test 32 blocks in a single test case, the probability of there being 1 false-negative is [~2.6%](https://www.wolframalpha.com/input/?i=%2832+choose+1%29+*+%2827%2F32863%29%5E1*%281-27%2F32863%29%5E%2832-1%29) and the probability of there being 2 false-negatives is [~0.03%](https://www.wolframalpha.com/input/?i=%2832+choose+2%29+*+%2827%2F32863%29%5E2*%281-27%2F32863%29%5E%2832-2%29).

2.6% is a little high, I agree, so I'll update the test case to look at the output and check how many failures there were in case of a non-zero exit code, only failing if there was more than 2.

Reply:
Just as a note, I ended up actually testing 1023 blocks (not just 32) because I fixed a separate problem that allowed me to actually test the expected number of blocks (had to specify `iflag=fullblock` in the dd command). So re-running the calculations I went with a failure tolerance of 4 blocks as [the chances of 5 block failures drops down to ~0.1%](https://www.wolframalpha.com/input/?i=%281023+choose+5%29+*+%2827%2F32863%29%5E5*%281-27%2F32863%29%5E%281023-5%29)